### PR TITLE
SQL: Fix null handling for IN => painless script

### LINKED ISSUE #35122: SQL: IN translated to painless has issue with null handling
When `null` literal is included in the list of values of the `IN` expression then if there is no match the expression should evaluate to `null`. It is the responsibility of `nullSafeFilter()` which wraps the whole filter to transform the `null` to false.

Currently the generation of the painless script skips the null values so it will not evaluate to `null` but directly to false. This can be problematic if the `IN` expression is nested under an `AND` or `OR` expression which requires the `null` to be returned in order to correctly evaluate itself using the 3-valued logic.

### PR DESCRIPTION
**Pull request by matriv**
Include `null` literals when generating the painless script for `IN` expressions.
Previously, they were skipped, because of an issue that has been fixed with #35108.

Fixes: #35122

### CODE REVIEW

**Review by astefan — APPROVED** [review]
LGTM

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java [hunk 1]
```diff
@@ -84,11 +84,12 @@ public boolean foldable() {
 
     @Override
     public Boolean fold() {
+        // Optimization for early return and Query folding to LocalExec
         if (value.dataType() == DataType.NULL) {
             return null;
         }
         if (list.size() == 1 && list.get(0).dataType() == DataType.NULL) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java]: What does this line do?

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java [hunk 2]
```diff
@@ -128,18 +129,15 @@ public ScriptTemplate asScript() {
         String scriptPrefix = leftScript + "==";
         LinkedHashSet<Object> values = list.stream().map(Expression::fold).collect(Collectors.toCollection(LinkedHashSet::new));
         for (Object valueFromList : values) {
-            // if checked against null => false
-            if (valueFromList != null) {
-                if (valueFromList instanceof Expression) {
-                    ScriptTemplate rightScript = asScript((Expression) valueFromList);
-                    sj.add(scriptPrefix + rightScript.template());
-                    rightParams.add(rightScript.params());
+            if (valueFromList instanceof Expression) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java]: I don't think this block ever executing considering the values is a list of _folded_ expressions and thus values.

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java [hunk 3]
```diff
@@ -128,18 +129,15 @@ public ScriptTemplate asScript() {
...
+                ScriptTemplate rightScript = asScript((Expression) valueFromList);
+                sj.add(scriptPrefix + rightScript.template());
+                rightParams.add(rightScript.params());
+            } else {
+                if (valueFromList instanceof String) {
```
Comment (costin) [inline x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java]: The script should just delegate to `InternalSqlScriptsUtils.in` not duplicate the code inside the script.

File: x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java [hunk 4]
```diff
@@ -254,7 +255,8 @@ public void testTranslateInExpression_HavingClause_PainlessAndNullHandling() {
         QueryTranslation translation = QueryTranslator.toQuery(condition, true);
         assertNull(translation.query);
         AggFilter aggFilter = translation.aggFilter;
-        assertEquals("InternalSqlScriptUtils.nullSafeFilter(params.a0==10 || params.a0==20 || params.a0==30)",
+        assertEquals("InternalSqlScriptUtils.nullSafeFilter(" +
```
Comment (costin) [inline x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java]: This should be `nullSafeFilter(InternalSqlScriptUtils.in(params.a0, params)` with the constants being sent as a params to the script

**Review by costin — CHANGES_REQUESTED** [review]
The script generation should delegate to InternalSqlScriptUtils not replicate the code.

**Comment by costin** [comment]
I realize this is impacted by #35055 but it makes the change set confusing. Could you please rebase this over that or potentially merge that first and the update this ticket.
It will be useful also in the future when looking back and what the changes actually were (looks to be the return of `null` ?).

Thanks

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java (hunk 1 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java]: it's the same optimization to fold to LocalExec. I can merge the ifs

File: x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java (hunk 3 above)
Comment (matriv) [reply x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java]: This will be addressed in the upcoming PR: #35055

**Comment by matriv** [comment]
OK! Will merge the other PR first and adjust this based on the other.

**Comment by matriv** [comment]
@costin updated on top of the #35055 and the fix is just to remove the null filtering: https://github.com/elastic/elasticsearch/pull/35124/files#diff-c67de98b2f74bb15fb8cdcbea8ef573cL112

**Comment by matriv** [comment]
Backported to `6.x` with bdf5d0581ac7ff200a77086f9228eb75371d4dca

**Pull request merged**